### PR TITLE
Expansion of class QualityControlReport

### PR DESCRIPTION
**Pull request by aclum**
- #1333 

Example data for expanding classes and slots for quality_control_report

### CODE REVIEW

**Comment by aclum** [comment]
Note tests are expected to fail as the actual schema hasn't been changed yet, I've only added example data to facilitate discussion.

**Review by Michal-Babins — COMMENTED** [review]
This looks like a straight forward way to address workflow state in the case where we cannot run it to completion. 

It looks at: status, what caused failure to occur, what workflow activity had the failure, and assumingely some extracted text from stderr. 

The question that arrises is, does this require and updated status for success as well? 

Also we see errors that do not address the direct cause of failure, how do we proceed in dealing with those? What would be an acceptable "comment"?

**Comment by turbomam** [comment]
Will the `failure_where` range be constrained (technically or policy-wise) to nmdc-schema class names?  `nmdc:wfrqc-11-hemh0a87.1`  is a `nmdc:ReadQcAnalysisActivity` , so what is the added benefit of specifying `failure_where: ReadQcAnalysisActivity` ?

**Comment by aclum** [comment]
@turbomam failure_where, at least how JGI uses is the process area that is the root cause of the failure. With my proposal the range would be either a PlannedProcess or a WorkflowExecutionActivity. An example of when the failure_where would be something other than the WorkflowExecutionActivity subclass is nmdc:wfrqc-11-c3eg9v51.1, running  ReadQcAnalysisActivity removed 99% of the data for hitting artifact sequences. This is a issue with the library preparation process so the failure_where would be LibraryPreparation. Yes, there will be an enumeration for failure_what.

We discussed some of Michal's in some meetings. quality_control_status would only be required if has_output did not exist or was an empty list. As we start this we should review the errors before submitting failures automatically to the schema. Having qc criteria after each workflow should cut down on erroneous error modes.

**Comment by scanon** [comment]
I wonder if we should include some provenance information to know what quality check/version was applied?

**Comment by aclum** [comment]
@scanon if we include the threshold files in the workflow repos it should be covered by the existing git_url and version slots.

**Comment by turbomam** [comment]
> Can you help me fix range for failure_where... I am trying to set the allowable range for failure_where to be any NMDC PlannedProcess Class or WorkflowExecutuionActivity Class.

I will have to chat with @cmungall or do some experimentation. I realize now that I know how to require that a slot takes its values from the identifiers of classes in the schema, **if the slot is a [type designator](https://linkml.io/linkml-model/latest/docs/designates_type/).** But I don't think that's the pattern we want to follow here.

**Comment by turbomam** [comment]
I raised this question in the [linkml-developers channel of the Monarch Initiative Slack](https://monarchinitiative.slack.com/archives/C01LDTVURH9/p1700579368753549)

File: src/data/valid/ReadQcAnalysisActivity-quality_fail.yaml
Comment (mbthornton-lbl) [inline src/data/valid/ReadQcAnalysisActivity-quality_fail.yaml:24]: suggest comment: valid passing activity

File: src/data/valid/ReadQcAnalysisActivity-quality_fail.yaml
Comment (mbthornton-lbl) [inline src/data/valid/ReadQcAnalysisActivity-quality_fail.yaml:2]: suggest comment: valid failing activity

File: src/schema/nmdc.yaml
Comment (mbthornton-lbl) [inline src/schema/nmdc.yaml:2183]: Add slot for Other

**Comment by turbomam** [comment]
it looks like the "no data generated" PV doesn't really mean **no** data generated

**Review by mbthornton-lbl — CHANGES_REQUESTED** [review]
Add Other failure category and a couple of explanatory comments to invalid testing data

**Comment by Michal-Babins** [comment]
> 

Maybe something like "No valid data generated" which would indicate there is some "valid" metric we apply.

**Comment by turbomam** [comment]
the `failure_categorization` slot should assert its `range` as the  `FailureCategorization` class

File: src/schema/nmdc.yaml
Comment (cmungall) [inline src/schema/nmdc.yaml:2285]: It looks like these are shadowing classes - we don't have an easy way to auto-shadow but I'd recommend using the exact same names for consistency, e.g. OmicsProcessing.

Comment (aclum) [reply src/schema/nmdc.yaml:2285]: @cmungall  I was following the style guide for permissible values, I will update to the same names if that is your suggestion. I did have a question about if a Class and permissible value can share the same name but based on yoru comment this is allowed.

**Comment by turbomam** [comment]
I don't belive that shared names are allowed in OWL. I would like to test
that first before we merge anything.

**Comment by turbomam** [comment]
I have tested the sharing of names between classes and permissible values. My findings are documented in the issue comments below.
- https://github.com/microbiomedata/nmdc-schema/issues/1419

@aclum this one specific case of name sharing is acceptable to me. We can work on whatever other issues this PR has and then merge it. Thanks for your patience.

**Comment by turbomam** [comment]
Current failure:

```shell
poetry run linkml-run-examples \
        --schema project/nmdc_schema_generated.yaml \
        --input-directory src/data/valid \
        --counter-example-input-directory src/data/invalid \
        --output-directory examples/output > examples/output/README.md
```

> Traceback (most recent call last):
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/bin/linkml-run-examples", line 8, in <module>
>     sys.exit(cli())
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/click/core.py", line 1157, in __call__
>     return self.main(*args, **kwargs)
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/click/core.py", line 1078, in main
>     rv = self.invoke(ctx)
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/click/core.py", line 1434, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/click/core.py", line 783, in invoke
>     return __callback(*args, **kwargs)
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml/workspaces/example_runner.py", line 307, in cli
>     runner.process_examples()
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml/workspaces/example_runner.py", line 137, in process_examples
>     self.process_examples_from_list(input_examples, fmt, False)
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml/workspaces/example_runner.py", line 161, in process_examples_from_list
>     validator = self.validator
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml/workspaces/example_runner.py", line 111, in validator
>     self._validator = JsonSchemaDataValidator(self.schemaview.schema)
>   File "<string>", line 3, in __init__
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml/validators/jsonschemavalidator.py", line 59, in __setattr__
>     self._hashable_schema = HashableSchemaDefinition(**asdict(__value))
>   File "<string>", line 66, in __init__
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml_runtime/linkml_model/meta.py", line 593, in __post_init__
>     self._normalize_inlined_as_dict(slot_name="classes", slot_type=ClassDefinition, key_name="name", keyed=True)
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml_runtime/utils/yamlutils.py", line 100, in _normalize_inlined_as_dict
>     self._normalize_inlined(slot_name, slot_type, key_name, keyed, False)
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml_runtime/utils/yamlutils.py", line 205, in _normalize_inlined
>     form_1({k: v})
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml_runtime/utils/yamlutils.py", line 153, in form_1
>     order_up(key, slot_type(**as_dict(raw_obj)))
>   File "<string>", line 72, in __init__
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml_runtime/linkml_model/meta.py", line 2476, in __post_init__
>     self.rules = [v if isinstance(v, ClassRule) else ClassRule(**as_dict(v)) for v in self.rules]
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml_runtime/linkml_model/meta.py", line 2476, in <listcomp>
>     self.rules = [v if isinstance(v, ClassRule) else ClassRule(**as_dict(v)) for v in self.rules]
>   File "<string>", line 44, in __init__
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml_runtime/linkml_model/meta.py", line 2595, in __post_init__
>     self.preconditions = AnonymousClassExpression(**as_dict(self.preconditions))
>   File "<string>", line 44, in __init__
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml_runtime/linkml_model/meta.py", line 2399, in __post_init__
>     self._normalize_inlined_as_dict(slot_name="slot_conditions", slot_type=SlotDefinition, key_name="name", keyed=True)
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml_runtime/utils/yamlutils.py", line 100, in _normalize_inlined_as_dict
>     self._normalize_inlined(slot_name, slot_type, key_name, keyed, False)
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml_runtime/utils/yamlutils.py", line 205, in _normalize_inlined
>     form_1({k: v})
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml_runtime/utils/yamlutils.py", line 153, in form_1
>     order_up(key, slot_type(**as_dict(raw_obj)))
>   File "<string>", line 115, in __init__
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml_runtime/linkml_model/meta.py", line 2277, in __post_init__
>     self.value_presence = PresenceEnum(self.value_presence)
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml_runtime/utils/enumerations.py", line 46, in __init__
>     if key not in self.__class__ and self._defn.code_set:
>   File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/linkml_runtime/utils/enumerations.py", line 28, in __contains__
>     return item in cls.__dict__
> TypeError: unhashable type: 'JsonObj'
> make: *** [project.Makefile:277: examples/output] Error 1

File: src/schema/nmdc.yaml
Comment (turbomam) [inline src/schema/nmdc.yaml:252]: In general, I hope that we can decrease the number of top level classes, not increase them. But I don't have a good parent class for this in mind right now.

**Comment by turbomam** [comment]
```shell
poetry run linkml-run-examples \
        --schema project/nmdc_schema_generated.yaml \
        --input-directory src/data/valid \
        --counter-example-input-directory src/data/invalid \
        --output-directory examples/output > examples/output/README.md

```

File: src/schema/nmdc.yaml
Comment (aclum) [reply src/schema/nmdc.yaml]: Permissible values have been updated to match the class names.

File: src/schema/nmdc.yaml
Comment (aclum) [reply src/schema/nmdc.yaml:253]: noted

**Comment by aclum** [comment]
Closing this PR since Mark's been working out of this pull request: https://github.com/microbiomedata/nmdc-schema/pull/1428

**Comment by aclum** [comment]
@turbomam  I deleted this branch just FYI in favor of yours to avoid confusion.

**Pull request closed without merging**
